File: store-service/internal/packaging/packaging_test.go
package packaging

import (
	"slices"
	"testing"
)

func TestIsValidShippingMode(t *testing.T) {
	cases := []struct {
		mode  ShippingMode
		valid bool
	}{
		{Air, true},
		{Land, true},
		{Sea, true},
		{"rocket", false},
		{"", false},
		{"AIR", false}, // case-sensitive — lowercase only per the spec
	}
	for _, c := range cases {
		t.Run(string(c.mode), func(t *testing.T) {
			if got := IsValidShippingMode(c.mode); got != c.valid {
				t.Errorf("IsValidShippingMode(%q) = %v, want %v", c.mode, got, c.valid)
			}
		})
	}
}

func TestShippingModes(t *testing.T) {
	modes := ShippingModes()
	want := []ShippingMode{Air, Land, Sea}
	if !slices.Equal(modes, want) {
		t.Errorf("ShippingModes() = %v, want %v", modes, want)
	}
}

func TestBuild_MaterialBySize(t *testing.T) {
	cases := []struct {
		size     Size
		expected Material
	}{
		{XLarge, Wood},
		{Large, Wood},
		{Medium, Cardboard},
		{Small, Plastic},
		{XSmall, Plastic},
	}
	for _, c := range cases {
		t.Run(string(c.size), func(t *testing.T) {
			pkg, err := Build(c.size, Air)
			if err != nil {
				t.Fatalf("Build(%q, Air) unexpected error: %v", c.size, err)
			}
			if got := pkg.Material(); got != c.expected {
				t.Errorf("Build(%q, Air).Material() = %q, want %q", c.size, got, c.expected)
			}
		})
	}
}

func TestBuild_ProtectionsByModeAndMaterial(t *testing.T) {
	cases := []struct {
		name     string
		size     Size
		mode     ShippingMode
		expected []Protection
	}{
		{"air_wood", Large, Air, []Protection{Polystyrene}},
		{"air_cardboard", Medium, Air, []Protection{Polystyrene}},
		{"air_plastic", Small, Air, []Protection{BubbleWrap}},
		{"land_wood", Large, Land, []Protection{Polystyrene}},
		{"land_cardboard", Medium, Land, []Protection{Polystyrene}},
		{"land_plastic", Small, Land, []Protection{Polystyrene}},
		{"sea_wood", Large, Sea, []Protection{MoistureBeads, BubbleWrap}},
		{"sea_cardboard", Medium, Sea, []Protection{MoistureBeads, BubbleWrap}},
		{"sea_plastic", Small, Sea, []Protection{MoistureBeads, BubbleWrap}},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			pkg, err := Build(c.size, c.mode)
			if err != nil {
				t.Fatalf("Build(%q, %q) unexpected error: %v", c.size, c.mode, err)
			}
			if got := pkg.Protections(); !slices.Equal(got, c.expected) {
				t.Errorf("Build(%q, %q).Protections() = %v, want %v", c.size, c.mode, got, c.expected)
			}
		})
	}
}

func TestBuild_UnknownShippingMode(t *testing.T) {
	cases := []ShippingMode{"rocket", "", "AIR" /* wrong case */}
	for _, mode := range cases {
		t.Run(string(mode), func(t *testing.T) {
			pkg, err := Build(Large, mode)
			if err == nil {
				t.Errorf("Build(Large, %q) returned nil error for unknown shipping mode", mode)
			}
			if pkg.Material() != "" || len(pkg.Protections()) != 0 {
				t.Errorf("Build(Large, %q) returned non-zero Package on error: %+v", mode, pkg)
			}
		})
	}
}

func TestBuild_UnknownSize(t *testing.T) {
	cases := []Size{"Huge", "", "XXLarge", "large" /* wrong case */}
	for _, size := range cases {
		t.Run(string(size), func(t *testing.T) {
			pkg, err := Build(size, Air)
			if err == nil {
				t.Errorf("Build(%q, Air) returned nil error for unknown size", size)
			}
			// Zero-value Package guarantee — callers that ignore the error
			// shouldn't get a populated package.
			if pkg.Material() != "" || len(pkg.Protections()) != 0 {
				t.Errorf("Build(%q, Air) returned non-zero Package on error: %+v", size, pkg)
			}
		})
	}
}

// The service is a thin adapter over Build; one test that round-trips a
// happy-path and an error-path is enough to catch regressions where the
// method signature drifts from the underlying function.
func TestPackagingService_Build(t *testing.T) {
	svc := NewService()
	if svc.Name() != "packaging" {
		t.Errorf("Name() = %q, want %q", svc.Name(), "packaging")
	}

	pkg, err := svc.Build(Large, Air)
	if err != nil {
		t.Fatalf("Build(Large, Air): %v", err)
	}
	if pkg.Material() != Wood {
		t.Errorf("Material = %q, want %q", pkg.Material(), Wood)
	}

	if _, err := svc.Build("Humongous", Air); err == nil {
		t.Error("Build(Humongous, Air): expected error on unknown size")
	}
}

File: backend/src/packaging/packaging.test.js
import { describe, it, expect } from "vitest";
import {
  build,
  isValidShippingMode,
  shippingModes,
  MATERIALS,
  PROTECTIONS,
  SIZES,
  SHIPPING_MODES,
} from "./packaging.js";
import { PackagingService } from "./service.js";

describe("isValidShippingMode", () => {
  it.each([
    [SHIPPING_MODES.Air, true],
    [SHIPPING_MODES.Land, true],
    [SHIPPING_MODES.Sea, true],
    ["rocket", false],
    ["", false],
    ["AIR", false], // case-sensitive — lowercase only
  ])("%s → %s", (mode, expected) => {
    expect(isValidShippingMode(mode)).toBe(expected);
  });
});

describe("shippingModes", () => {
  it("returns the canonical list in declaration order", () => {
    expect(shippingModes()).toEqual(["air", "land", "sea"]);
  });
  it("returns a fresh copy each time — mutating the result doesn't affect the registry", () => {
    const a = shippingModes();
    a.push("rocket");
    expect(shippingModes()).toEqual(["air", "land", "sea"]);
  });
});

describe("build — strategy pattern (size → material)", () => {
  it.each([
    [SIZES.XLarge, MATERIALS.Wood],
    [SIZES.Large,  MATERIALS.Wood],
    [SIZES.Medium, MATERIALS.Cardboard],
    [SIZES.Small,  MATERIALS.Plastic],
    [SIZES.XSmall, MATERIALS.Plastic],
  ])("%s → %s", (size, material) => {
    expect(build(size, SHIPPING_MODES.Air).material).toBe(material);
  });

  it("throws on unknown size", () => {
    expect(() => build("Humongous", SHIPPING_MODES.Air)).toThrow(/unknown size/);
  });
});

describe("build — decorator pattern (material + mode → protections)", () => {
  it.each([
    [SIZES.Large,  SHIPPING_MODES.Air,  [PROTECTIONS.Polystyrene]],   // wood + air
    [SIZES.Medium, SHIPPING_MODES.Air,  [PROTECTIONS.Polystyrene]],   // cardboard + air
    [SIZES.Small,  SHIPPING_MODES.Air,  [PROTECTIONS.BubbleWrap]],    // plastic + air
    [SIZES.Large,  SHIPPING_MODES.Land, [PROTECTIONS.Polystyrene]],   // land always polystyrene
    [SIZES.Small,  SHIPPING_MODES.Land, [PROTECTIONS.Polystyrene]],
    [SIZES.Large,  SHIPPING_MODES.Sea,  [PROTECTIONS.MoistureBeads, PROTECTIONS.BubbleWrap]], // sea = 2
    [SIZES.XSmall, SHIPPING_MODES.Sea,  [PROTECTIONS.MoistureBeads, PROTECTIONS.BubbleWrap]],
  ])("%s + %s → %j", (size, mode, expected) => {
    expect(build(size, mode).protections).toEqual(expected);
  });

  it("throws on unknown mode", () => {
    expect(() => build(SIZES.Large, "rocket")).toThrow(/unknown shipping mode/);
  });
});

describe("Package is frozen", () => {
  it("doesn't allow mutation of material/protections", () => {
    const pkg = build(SIZES.Large, SHIPPING_MODES.Air);
    expect(() => { pkg.material = "gold"; }).toThrow();
    expect(() => { pkg.protections.push("gold"); }).toThrow();
  });
});

describe("PackagingService", () => {
  it("exposes build and delegates to the factory", () => {
    const svc = new PackagingService();
    const pkg = svc.build(SIZES.Large, SHIPPING_MODES.Air);
    expect(pkg.material).toBe(MATERIALS.Wood);
    expect(pkg.protections).toEqual([PROTECTIONS.Polystyrene]);
    expect(svc.entityName).toBe("packaging");
  });

  it("propagates errors on unknown size", () => {
    const svc = new PackagingService();
    expect(() => svc.build("Humongous", SHIPPING_MODES.Air)).toThrow(/unknown size/);
  });
});
